docs/blueprints/storage-node-downtime-tracking-with-audits.md

```diff
@@ -0,0 +1,153 @@
+
+## Implementation
+
+### 1) Determine the business requirement for the minimum frequency of audits per node. Implement the necessary changes to make this happen. 
+
+This will determine what window sizes we can work with.
+
+### 2) Add new columns, `offline_scored_at` and `offline_suspended`, and `under_review` to `nodes` table
+We do not need to evaluate a node's standing every time it is audited because we are only interested in complete windows. Therefore, we only need to evaluate a node's standing once per window. If `offline_scored_at` does not fall within the current window then we need to set `offline_scored_at` to the current time and evaluate whether the node needs to be suspended, reinstated, or disqualified.
+
+`offline_scored_at` and `under_review` should be timestamps. `offline_suspended` could be a timestamp or a boolean. The boolean would use less space, but the timestamp might be more informative on a SNO dashboard.
+
+### 3) Implement a DB table to store audit windows
+
+```sql
+CREATE TABLE audit_windows (
+    node_id BYTEA,
+    offline INT,
+    total INT,
+    window_start TIMESTAMP,
+)
+```
+The `window_start` column refers to the start boundary of the window. This can be determined by truncating the current time down to the nearest multiple of the window size.
+
+### 4) Add `AuditWindowConfig` to overlay config. 
+
+```
+type AuditWindowConfig struct {
+    WindowSize       time.Duration `help:"the length of time spanning a single audit window."`
+    TrackingPeriod   time.Duration `help:"the length of time to track audit windows for node suspension and disqualification."`
+    GracePeriod      time.Duration `help:"The length of time to give suspended SNOs to diagnose and fix issues causing downtime. Afterwards, they will have one tracking period to reach the minimum online score before disqualification."`
+    OfflineThreshold float64       `help:"The point above which a node is punished for offline audits. Determined by calculating the percentage of offline audits within each window and finding the average across windows within the tracking period."`
+}
+```
+
+### 5) Write to audit windows and evaluate node standing. 
+The overlay DB method `BatchUpdateStats` is where a node's audit reputation is updated by audit results. It iterates through each node, retrieving its dossier from the `nodes` table and updates its reputation. We also need to update and read information from the node dossier to set and evaluate a node's standing regarding audit windows, so this method is a good place to do this. 
+
+Add a new argument to `BatchUpdateStats` to pass in `AuditWindowConfig`. This gives us the window size parameter required to write to the audit windows table and the tracking period, grace period, and offline threshold parameters required to evaluate whether a node needs to be suspended, reinstated, or disqualified.
```

Comment:
it might be worth mentioning that this will probably have a (hopefully small) performance impact on audits. It depends on how often the audit_windows query has to run per audit